src/passes/StringLowering.cpp

```diff
@@ -0,0 +1,188 @@
+      StringPtrs& stringPtrs;
+
+      StringWalker(StringPtrs& stringPtrs) : stringPtrs(stringPtrs) {}
+
+      void visitStringConst(StringConst* curr) {
+        stringPtrs.push_back(getCurrentPointer());
+      }
+    };
+
+    ModuleUtils::ParallelFunctionAnalysis<StringPtrs> analysis(
+      *module, [&](Function* func, StringPtrs& stringPtrs) {
+        if (!func->imported()) {
+          StringWalker(stringPtrs).walk(func->body);
+        }
+      });
+
+    // Also walk the global module code (for simplicity, also add it to the
+    // function map, using a "function" key of nullptr).
+    auto& globalStrings = analysis.map[nullptr];
+    StringWalker(globalStrings).walkModuleCode(module);
+
+    // Combine all the strings.
+    std::unordered_set<Name> stringSet;
+    for (auto& [_, currStringPtrs] : analysis.map) {
+      for (auto** stringPtr : currStringPtrs) {
+        stringSet.insert((*stringPtr)->cast<StringConst>()->string);
+        stringPtrs.push_back(stringPtr);
+      }
+    }
+
+    // Generate the indexes from the combined set of necessary strings, which we
+    // sort for determinism (alphabetically).
+    for (auto& string : stringSet) {
+      strings.push_back(string);
+    }
+    std::sort(strings.begin(), strings.end());
+    for (Index i = 0; i < strings.size(); i++) {
+      stringIndexes[strings[i]] = i;
+    }
+  }
+
+  // For each string index, the name of the global that replaces it.
+  std::vector<Name> globalNames;
+
+  Type nnstringref = Type(HeapType::string, NonNullable);
+
+  // Existing globals already in the form we emit can be reused. That is, if
+  // we see
+  //
+  //  (global $foo (ref null string) (string.const ..))
+  //
+  // then we can just use that as the global for that string. This avoids
+  // repeated executions of the pass adding more and more globals.
+  //
+  // Note that we don't note these in newNames: They are already in the right
+  // sorted position, before any uses, as we use the first of them for each
+  // string. Only actually new names need sorting.
+  //
+  // Any time we reuse a global, we must not modify its body (or else we'd
+  // replace the global that all others read from); we note them here and
+  // avoid them in replaceStrings later to avoid such trampling.
+  std::unordered_set<Expression**> stringPtrsToPreserve;
+
+  void addGlobals(Module* module) {
+    // Note all the new names we create for the sorting later.
+    std::unordered_set<Name> newNames;
+
+    // Each string will get a global.
+    globalNames.resize(strings.size());
+
+    // Find globals to reuse (see comment on stringPtrsToPreserve for context).
+    for (auto& global : module->globals) {
+      if (global->type == nnstringref && !global->imported()) {
+        if (auto* stringConst = global->init->dynCast<StringConst>()) {
+          auto stringIndex = stringIndexes[stringConst->string];
+          auto& globalName = globalNames[stringIndex];
```

Comment:
It looks like the only use of `stringIndexes` is to look up an index into `globalNames`. Can we skip a step here and have `globalNames` map directly from string to name?

Reply:
Good idea! Done.